src/miner.cpp

```diff
@@ -351,332 +385,387 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const MinerAddre
+    const int64_t nMedianTimePast = pindexPrev->GetMedianTimePast();
+    CCoinsViewCache view(pcoinsTip);
+
+    SaplingMerkleTree sapling_tree;
+    assert(view.GetSaplingAnchorAt(view.GetBestAnchor(SAPLING), sapling_tree));
+
+    nLockTimeCutoff = (STANDARD_LOCKTIME_VERIFY_FLAGS & LOCKTIME_MEDIAN_TIME_PAST)
+                       ? nMedianTimePast
+                       : pblock->GetBlockTime();
+
+    // We want to track the value pool, but if the miner gets
+    // invoked on an old block before the hardcoded fallback
+    // is active we don't want to trip up any assertions. So,
+    // we only adhere to the turnstile (as a miner) if we
+    // actually have all of the information necessary to do
+    // so.
+    if (chainparams.ZIP209Enabled()) {
+        if (pindexPrev->nChainSproutValue) {
+            sproutValue = *pindexPrev->nChainSproutValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+        if (pindexPrev->nChainSaplingValue) {
+            saplingValue = *pindexPrev->nChainSaplingValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+        if (pindexPrev->nChainOrchardValue) {
+            orchardValue = *pindexPrev->nChainOrchardValue;
+        } else {
+            monitoring_pool_balances = false;
+        }
+    }
 
-    // Collect memory pool transactions into the block
-    CTxMemPool::setEntries inBlock;
-    CTxMemPool::setEntries waitSet;
+    addPriorityTxs();
+    addScoreTxs();
 
-    // This vector will be sorted into a priority queue:
-    vector<TxCoinAgePriority> vecPriority;
-    TxCoinAgePriorityCompare pricomparer;
-    std::map<CTxMemPool::txiter, double, CTxMemPool::CompareIteratorByHash> waitPriMap;
-    typedef std::map<CTxMemPool::txiter, double, CTxMemPool::CompareIteratorByHash>::iterator waitPriIter;
-    double actualPriority = -1;
+    last_block_num_txs = nBlockTx;
+    last_block_size = nBlockSize;
+    LogPrintf("%s: total size %u (excluding coinbase) txs: %u fees: %ld sigops %d", __func__, nBlockSize, nBlockTx, nFees, nBlockSigOps);
 
-    std::priority_queue<CTxMemPool::txiter, std::vector<CTxMemPool::txiter>, ScoreCompare> clearedTxs;
-    bool fPrintPriority = GetBoolArg("-printpriority", DEFAULT_PRINTPRIORITY);
-    uint64_t nBlockSize = 1000;
-    uint64_t nBlockTx = 0;
-    unsigned int nBlockSigOps = 100;
-    int lastFewTxs = 0;
-    CAmount nFees = 0;
+    // Create coinbase tx
+    if (next_cb_mtx) {
+        pblock->vtx[0] = *next_cb_mtx;
+    } else {
+        pblock->vtx[0] = CreateCoinbaseTransaction(chainparams, nFees, minerAddress, nHeight);
+    }
+    pblocktemplate->vTxFees[0] = -nFees;
+
+    // Update the Sapling commitment tree.
+    for (const CTransaction& tx : pblock->vtx) {
+        for (const OutputDescription& odesc : tx.vShieldedOutput) {
+            sapling_tree.append(odesc.cmu);
+        }
+    }
+
+    // Randomise nonce
+    arith_uint256 nonce = UintToArith256(GetRandHash());
```

Comment:
I'm surprised that selecting a nonce is part of constructing the block template. I guess this is just so that there's *something* there?